Services.Test/CustomDeviceModelsTest.cs

```diff
@@ -0,0 +1,188 @@
+        public void InitialListIsEmpty()
+        {
+            // Arrange
+            this.ThereAreNoCustomDeviceModelsInTheStorage();
+
+            // Act
+            var result = this.target.GetListAsync().Result;
+
+            // Assert
+            Assert.Empty(result);
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void CreateCustomDeviceModel()
+        {
+            // Arrange
+            var deviceModel = new DeviceModel { ETag = "Etag_1" };
+            this.UpdateDeviceModelInStorage(deviceModel);
+
+            // Act
+            DeviceModel result = this.target.InsertAsync(deviceModel).Result;
+
+            // Assert
+            Assert.NotNull(result);
+            Assert.Equal(deviceModel.ETag, result.ETag);
+        }
+
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void CreatedCustomDeviceModelsAreStored()
+        {
+            // Arrange
+            var deviceModel = new DeviceModel { ETag = "Etag_1" };
+            this.UpdateDeviceModelInStorage(deviceModel);
+
+            // Act
+            this.target.InsertAsync(deviceModel).Wait();
+
+            // Assert
+            this.storage.Verify(x => x.UpdateAsync(
+                STORAGE_COLLECTION,
+                It.IsAny<string>(),
+                It.IsAny<string>(),
+                null));
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void CustomDeviceModelsCanBeUpserted()
+        {
+            // Arrange
+            var deviceModel = new DeviceModel { ETag = "oldEtag" };
+            var updatedDeviceModel = new DeviceModel  { ETag = "newETag" };
+
+            this.SaveDeviceModelInStorage(deviceModel);
+            this.UpdateDeviceModelInStorage(updatedDeviceModel);
+
+            // Act
+            this.target.UpsertAsync(deviceModel).Wait();
+
+            // Assert
+            this.storage.Verify(x => x.UpdateAsync(
+                STORAGE_COLLECTION,
+                It.IsAny<string>(),
+                It.IsAny<string>(),
+                "oldEtag"));
+
+            Assert.Equal("newETag", deviceModel.ETag);
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void CreateNewCustomDeviceModelsIfNotExisted()
+        {
+            // Arrange
+            var deviceModel = new DeviceModel { Id = "id", ETag = "Etag" };
+
+            this.ThereIsNoCustomDeviceModelsByIdInTheStorage(deviceModel.Id);
+            this.UpdateDeviceModelInStorage(deviceModel);
+
+            // Act
+            this.target.UpsertAsync(deviceModel).Wait();
+
+            // Assert
+            this.storage.Verify(x => x.UpdateAsync(
+                STORAGE_COLLECTION,
+                It.IsAny<string>(),
+                It.IsAny<string>(),
+                null));
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void CustomDeviceModelsCanNotBeUpsertedIfEtagNotMatch()
```

Comment:
Nit: rename to incluse "DoesNot" CustomDeviceModelsCanNotBeUpsertedIfEtagDoesNotMatch()